### ISSUE TITLE
ping/pong timeout recursion

### ISSUE THREAD

**Issue by adamwojt** [websocket-client/websocket-client#858]
Discovered (thankfully in dev env) issue with ping/pong timeout when running with `run_forever`. Chain traceback, with multiple (10,000+) ping/pong timeout errors. At the end, our logger raised recursion error as it couldn't keep up. We are running with proxy and these parameters:
``` python
ping_timeout: int = 4
ping_interval: int = 5
``` 

``` log
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/websocket/_app.py", line 354, in setSock
    dispatcher.read(self.sock.sock, read, check)
  File "/usr/local/lib/python3.9/site-packages/websocket/_app.py", line 56, in read
    check_callback()
  File "/usr/local/lib/python3.9/site-packages/websocket/_app.py", line 397, in check
    raise WebSocketTimeoutException("ping/pong timed out")
websocket._exceptions.WebSocketTimeoutException: ping/pong timed out

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/websocket/_app.py", line 354, in setSock
    dispatcher.read(self.sock.sock, read, check)
  File "/usr/local/lib/python3.9/site-packages/websocket/_app.py", line 56, in read
    check_callback()
  File "/usr/local/lib/python3.9/site-packages/websocket/_app.py", line 397, in check
    raise WebSocketTimeoutException("ping/pong timed out")
websocket._exceptions.WebSocketTimeoutException: ping/pong timed out

During handling of the above exception, another exception occurred:

....
```

**Comment by bubbleboy14** [comment]
Hello @adamwojt . Could you please provide a minimal test case (client and server if possible) that reproduces the issue?

**Comment by adamwojt** [comment]
@bubbleboy14 

Sure here is how i was able to reproduce it locally with `1.4.0` (same as in our env at the time of issue occurring):

**Install**:
1. `pip install websocket-client==1.4.0`
2. `pip install websockets`

**client.py**
``` python
import websocket
import logging
logging.basicConfig()

websocket_url = "ws://localhost:8765"
ping_timeout = 1
ping_interval = 2

def on_message(ws, data):
    print(data)

def on_error(ws, error):
    logging.exception("error")

def on_open(ws):
    print("Connection opened")

ws = websocket.WebSocketApp(
    websocket_url,
    on_message=on_message,
    on_error=on_error,
    on_open=on_open,
)

ws.run_forever(
    ping_timeout=ping_timeout,
    ping_interval=ping_interval,
)
```

server.py
``` python
#!/usr/bin/env python
import asyncio
import websockets

async def hello(websocket, path):
    name = await websocket.recv()
    print("< {}".format(name))

    greeting = "Hello {}!".format(name)
    await websocket.send(greeting)
    print("> {}".format(greeting))

start_server = websockets.serve(hello, 'localhost', 8765)
print("Running in localhost:8765")

asyncio.get_event_loop().run_until_complete(start_server)
asyncio.get_event_loop().run_forever()
```

**Reproduce**:
1. Open two seperate shells
2. `python server.py` to run server
3. `python client.py` to run client
4. close `server.py` process and reopen it quickly.
5. Now `client.py` stuck in ping/pong timeout loop `During handling of the above exception, another exception occurred:`

The behavior seems to be different with `1.4.1` version as it doesn't seem to "run_forever" and `client.py` process exits as soon as `server.py` is closed. Quite confusing because changelog doesn't note this important change.

**Comment by bubbleboy14** [comment]
Awesome, thanx for the repro! Try this branch/pr @adamwojt:

https://github.com/websocket-client/websocket-client/pull/859

So WebSocketApp._send_ping() contains a while loop run in a thread. It looks like there is/was (on line 254) a "break" that basically ends the loop (of sending pings) if a ping fails (even after reconnect).

So this change should fix it. Does it work for you?

**Comment by engn33r** [comment]
I can recreate the  using release 1.4.0 and your example code. However, I did not encounter the recursive error when using the [long-lived connection example code](https://github.com/websocket-client/websocket-client#long-lived-connection). The long-lived example code specifies rel as a dispatcher and this prevents the recursion. @bubbleboy14 is the rel maintainer and might weigh in with why the recursive error happens when rel is not specified as a dispatcher. I agree the recursive error is a problem, but this can be resolved by following the long-lived connection example and specifying a dispatcher, so I do not think this is an error in the websocket-client code. If there was any other unexpected outcomes from your code other than the recursive error, please elaborate.

I was able to omit step 4 (restarting the server quickly) and the same error happens. To explain the scenario where the server is not restarted, my understanding is when the client is disconnected because the server is down, the ping/pong timeout does not stop immediately, because `ping_timeout` is designed to detect scenarios where the server isn't reachable within a certain amount of delay. Because the `ping_timeout` parameter determines when `run_forever` exits with a `WebSocketTimeoutException` exception ([as documented here](https://websocket-client.readthedocs.io/en/latest/examples.html#ping-pong-usage)), the disconnection causes the timeout to be reached and ends `run_forever`. I don't think the `reconnect` parameter passed to `run_forever` will override the ping/pong timeout disconnection.

The reason why this might happen even in the case where the server is restarted is that the ping/pong timeout system implemented is not "smart", it does not track multiple pings/pongs in transit. Instead websocket-client only retains awareness of when the last ping/pong happened. My detailed notes about this are found in #611.

**Comment by bubbleboy14** [comment]
Did anyone test https://github.com/websocket-client/websocket-client/pull/859 ?

As I recall, the "ping/pong timed out" in the traceback (provided by OP) was happening because the code stopped sending pings after the first reconnect (due to exiting the loop on ping failure).

The provided pull request (859) fixes this issue. Is there anything else (unaddressed) related to this ticket? Would anyone confirm whether this fix also works for them?

**Comment by j-a-n** [comment]
In my opinion #859 is not sufficient so resolve the issue for the following reasons:
- The ping Thread will continue to send pings after the websocket connection ist lost. This will result in further errors and reconnects.
- When using the default dispatcher, the DispatcherBase implementation will reconnect in a loop [_app.py#L55](https://github.com/websocket-client/websocket-client/blob/37429a3efa1238cb5b6e5c056dd4e9190575d5ca/websocket/_app.py#L55) while the `handleDisconnect` function will call the Dispatcher `reconnect` method again on every error ([_app.py#L435](https://github.com/websocket-client/websocket-client/blob/37429a3efa1238cb5b6e5c056dd4e9190575d5ca/websocket/_app.py#L435))
- The current reconnect implementation will end up in a `RecursionError: maximum recursion depth exceeded` after enought reconnects. This is because of the follwing recursion `setSock` -> `handleDisconnect` -> `DispatcherBase.reconnect` -> `setSock`

It would be nice if someone could test and review #862

**Comment by adamwojt** [comment]
@j-a-n @bubbleboy14 @engn33r
Actually putting ping/pong recursion error aside:

I think what mostly caused some headache in our project was that going from `1.3.3` > `1.4.0`  the re-connection mechanism was suddenly turned on by default which alone caused issues for us.

We were building our code assuming that `run_forever` will exit on ping pong timeout and then we were handling reconnects ourselves in custom loop. We also raise some errors explicitly in `on_error` as we are listening for some server-specific notification about end of session that require a fresh URL so it did make sense to have custom reconnect solution.

I think we briefly checked Changelog but it did not occur to us that automatic reconnect would be "default" behavior now as we knew from documentation that this required additional dependency (`rel`) that we did not want to introduce as simple custom loop was working for us just fine.

In `1.4.1` the reconnect doesn't happen automatically which made our code work as intended with our custom while loop.

**Comment by bubbleboy14** [comment]
> * The ping Thread will continue to send pings after the websocket connection ist lost. This will result in further errors and reconnects.

Hm, interesting. I haven't observed this, has anyone else? This certainly seems plausible.

> * When using the default dispatcher, the DispatcherBase implementation will reconnect in a loop [_app.py#L55](https://github.com/websocket-client/websocket-client/blob/37429a3efa1238cb5b6e5c056dd4e9190575d5ca/websocket/_app.py#L55) while the `handleDisconnect` function will call the Dispatcher `reconnect` method again on every error ([_app.py#L435](https://github.com/websocket-client/websocket-client/blob/37429a3efa1238cb5b6e5c056dd4e9190575d5ca/websocket/_app.py#L435))

The scenario described is actually the reason for the loop cited on line 55. The code avoids this (recursion spiral) by passing reconnecting=True on [line ](https://github.com/websocket-client/websocket-client/blob/master/websocket/_app.py#L58), which on https://github.com/websocket-client/websocket-client/blob/master/websocket/_app.py#L381 prevents the call to handleDisconnect().

> * The current reconnect implementation will end up in a `RecursionError: maximum recursion depth exceeded` after enought reconnects. This is because of the follwing recursion `setSock` -> `handleDisconnect` -> `DispatcherBase.reconnect` -> `setSock`

This doesn't occur for the reasons stated above.

> It would be nice if someone could test and review #862

Looks like a nice one!

**Comment by j-a-n** [comment]
Yes, you are right @bubbleboy14 , my description of the recursion part is not correct.
The looping reconnect is caused by the WebSocketTimeoutException in the read function.
After a disconnect of the socket, the ping thread continues (at least for while, more on that later).
So, one ping interval after the disconnect `last_ping_tm` will be set with no chance to receive a pong.
Then, after a reconnect of the websocket, the read function is called again and will throw a WebSocketTimeoutException.
The freshly connected websocket will be "reconnected" again.
After the next reconnect this will happen again and again.

Back to the ping thread.
The ping thread checks for `self.sock`, but `self.sock` is always true, checking for `self.sock.connected` would be better.
With #859 the ping thread will run forever (which is good) but it will constantly update `last_ping_tm` and try to send pings over a closed socket.
Without #859 the ping thread will stop after the first websocket disconnect and after a reconnect we are running without a ping thread.

Unfortunately, I can't remember how the infinite recursion happened.

**Comment by engn33r** [comment]
This should be fixed in v1.5.0 now that #862 has been merged
